src/grid-bots/multi-vendor-multi-contract.py

```diff
@@ -256,47 +255,128 @@ async def update_limit_orders(self):
                   'execInst': execInst,
+                  'targetAccountId' : self.VENUE_IDS[venue],
                   'text': 'Sent from ProfitView.net'
-                })
+                }
+              )
             except Exception as e:
               logger.error(e)
           else :
             # If I have a current open position in the opposite direction, double the order size
             multiplier = 1
-            if(self.VENUES[venue][sym]['current_risk'] <= 0):
-              multiplier = self.MULTIPLIER
-            if(self.VENUES[venue][sym]['direction'] == 'SHORT'):
+            reduce_only = False
+            if current_risk <= 0 :
+              multiplier = self.VENUES[venue][sym].get('multiplier', 1)
+            if self.VENUES[venue][sym]['direction'] == 'SHORT' :
+              if int(current_risk) == 0 :
+                bids=[]
               execInst = 'ParticipateDoNotInitiate,ReduceOnly'
+              reduce_only = True
             else:
               execInst = 'ParticipateDoNotInitiate'
-            for bid in bids:
-              try:
-                self.call_endpoint(
-                  venue,
-                  'order',
-                  'private',
-                  method = 'POST',
-                  params = {
-                    'symbol': sym,
-                    'side': 'Buy',
-                    'orderQty': self.VENUES[venue][sym]['grid_size'] * multiplier,
-                    'price': bid,
-                    'ordType': 'Limit',
-                    'execInst': execInst,
-                    'text': 'Sent from ProfitView.net'
-                  }
-                )
-              except Exception as e:
-                logger.error(e)
+
+            # First cycle, find all orders that already have the same price as bids
+            # * Use array copy to be able to remove found bids from original array
+            if bids :
+              for bid in bids[:]:
+                # We need available have orders
+                if orders_bid :
+                  # Find if there is already an order with same price
+                  order_id = self.order_exists(bid, orders_bid)
+                  if order_id :
+                    # The bid is present, we don't need in next cycles
+                    bids.remove(bid)
+                    # The order has been "consumed", remove from the temporary dict
+                    del orders_bid[order_id]
+                else :
+                  break
+
+            # Second cycle, amend existing orders as long as we have enough active orders
+            if bids :
+              for bid in bids[:]:
+                if orders_bid :
+                  # Use an order, and remove from list of available at same time
+                  order_id, _ = orders_bid.popitem()
+                  # Amend the order
+                  # self.amend_order(venue, order_id=order_id, price=bid)
+                  try:
+                    self.call_endpoint(
```

Comment:
imo, we won't HAVE to manage amends, we can still cancel and recreate the orders which will remove a lot of state management. you use only one more API request to cancel all so its not the end of the world to cancel / new

Reply:
I personally prefer the state management, it reduces calls by a lot, also amends are faster from an API standpoint because bitmex is column based so changing price happens faster then deleting an order ( with all its data ) and recreating it

If you prefer, I can add an option and have a path that uses the cancel all/recreate path and another that uses the state management